lib/hanami/helpers/form_helper/form_builder.rb

```diff
@@ -0,0 +1,1387 @@
+        #
+        # @see #password_field
+        EMPTY_STRING = ""
+        private_constant :EMPTY_STRING
+
+        include Hanami::View::Helpers::EscapeHelper
+        include Hanami::View::Helpers::TagHelper
+
+        # @api private
+        # @since 2.0.0
+        attr_reader :base_name
+        private :base_name
+
+        # @api private
+        # @since 2.0.0
+        attr_reader :inflector
+        private :inflector
+
+        # @api private
+        # @since 2.0.0
+        attr_reader :form_attributes
+        private :form_attributes
+
+        # Returns a new form builder.
+        #
+        # @param inflector [Dry::Inflector] the app inflector
+        # @param base_name [String, nil] the base name to use for all fields in the form
+        # @param values [Hanami::Helpers::FormHelper::Values] the values for the form
+        #
+        # @return [self]
+        #
+        # @see Hanami::Helpers::FormHelper#form_for
+        #
+        # @api private
+        # @since 2.0.0
+        def initialize(inflector:, form_attributes:, base_name: nil, values: Values.new)
+          @base_name = base_name
+          @values = values
+          @form_attributes = form_attributes
+          @inflector = inflector
+        end
+
+        # @api private
+        # @since 2.0.0
+        def call(content, **attributes)
+          attributes["accept-charset"] ||= DEFAULT_CHARSET
+
+          method_override, original_form_method = _form_method(attributes)
+          csrf_token, token = _csrf_token(@values, attributes)
+
+          tag.form(**attributes) do
+            (+"").tap { |inner|
+              inner << input(type: "hidden", name: "_method", value: original_form_method) if method_override
+              inner << input(type: "hidden", name: "_csrf_token", value: token) if csrf_token
+              inner << content
+            }.html_safe
+          end
+        end
+
+        # Applies the base input name to all fields within the given block.
+        #
+        # This can be helpful when generating a set of nested fields.
+        #
+        # This is a convenience only. You can achieve the same result by including the base name at
+        # the beginning of each input name.
+        #
+        # @param name [String] the base name to be used for all fields in the block
+        #
+        # @example Basic usage
+        #   <% f.fields_for "address" %>
+        #     <%= f.text_field "street" %>
+        #     <%= f.text_field "suburb" %>
+        #   <% end %>
+        #
+        #   # A convenience for:
+        #   # <%= f.text_field "address.street" %>
+        #   # <%= f.text_field "address.suburb" %>
+        #
+        #   =>
+        #   <input type="text" name="delivery[customer_name]" id="delivery-customer-name" value="">
+        #   <input type="text" name="delivery[address][street]" id="delivery-address-street" value="">
+        #
+        # @example Multiple levels of nesting
+        #   <% f.fields_for "address" %>
```

Comment:
@timriley To make the usage more explicit for the user, should we consider yielding a builder in the `fields_for` block?

Example:

```
<% f.fields_for "address" |fa| %>
  <%= fa.text_field "street" %>

  <% fa.fields_for "location" |fal| %>
    <%= fal.text_field "city" %>
  <% end %>
<% end %>
```

Reply:
Ah yeah, you're right, we should do this. I'll make sure I add this in before I merge this PR.